File: TestLeetCode/Tree/IsBalanced.cs
public class IsBalancedSolutionWithValueTuple {
    public bool IsBalanced(TreeNode root) {
        return DFS(root).IsBalanced;
    }

    private (bool IsBalanced, int Height) DFS(TreeNode root) {
        if (root == null) {
            return (true, 0);
        }

        var left = DFS(root.left);
        var right = DFS(root.right);
        bool isBalanced = left.IsBalanced && right.IsBalanced && Math.Abs(left.Height - right.Height) <= 1;

        return (isBalanced, 1 + Math.Max(left.Height, right.Height));
    }
}

public class IsBalancedSolution {
    public bool IsBalanced(TreeNode root) {
        return DFS(root).IsBalanced;
    }

    private BalanceStatus DFS(TreeNode root) {
        if (root == null) {
            return new BalanceStatus(true, 0);
        }

        BalanceStatus left = DFS(root.left);
        BalanceStatus right = DFS(root.right);
        bool balanced = left.IsBalanced && right.IsBalanced && Math.Abs(left.Height - right.Height) <= 1;

        return new BalanceStatus(balanced, 1 + Math.Max(left.Height, right.Height));
    }
}

public class BalanceStatus {
    public bool IsBalanced { get; }
    public int Height { get; }

    public BalanceStatus(bool isBalanced, int height) {
        IsBalanced = isBalanced;
        Height = height;
    }
}

File: python/tree/is_balanced.py
from python.common import TreeNode


class BalancedTree:
    def is_balanced(self, root: TreeNode | None) -> bool:
        def height(node):
            if not node: return 0
            left, right = height(node.left), height(node.right)
            if left < 0 or right < 0 or abs(left - right) > 1: return -1
            return 1 + max(left, right)
        return height(root) >= 0

    def is_balanced_with_status(self, root: TreeNode | None) -> bool: return self.is_balanced(root)
